src/plone/api/relation.py

```diff
@@ -0,0 +1,230 @@
+        raise InvalidParameterError('{} is no dexterity content'.format(source))
+
+    if not IDexterityContent.providedBy(target):
+        raise InvalidParameterError('{} is no dexterity content'.format(target))
+
+    if not isinstance(relationship, six.string_types):
+        raise InvalidParameterError('{} is no string'.format(relationship))
+
+    relation_catalog = getUtility(ICatalog)
+    intids = getUtility(IIntIds)
+    to_id = intids.getId(target)
+    from_id = intids.getId(source)
+    from_attribute = relationship
+
+    # Check if there is exactly this relation.
+    # If so remove it and create a fresh one.
+    query = {
+        'from_attribute': from_attribute,
+        'from_id': from_id,
+        'to_id': to_id,
+    }
+    for rel in relation_catalog.findRelations(query):
+        relation_catalog.unindex(rel)
+
+    if from_attribute == referencedRelationship:
+        # Don't mess with linkintegrity-relations!
+        # Refresh them by triggering this subscriber.
+        modifiedContent(source, None)
+        return
+
+    if ITERATE_RELATION_NAME is not None and from_attribute == ITERATE_RELATION_NAME:
+        # Iterate relations use a subclass of RelationValue
+        relation = StagingRelationValue(to_id)
+        event._setRelation(source, ITERATE_RELATION_NAME, relation)
+        return
+
+    fti = queryUtility(IDexterityFTI, name=source.portal_type)
+    if not fti:
+        logger.info(u'{} is no dexterity content'.format(source.portal_type))
+        return
+    field_and_schema = _get_field_and_schema_for_fieldname(from_attribute, fti)
+
+    if field_and_schema is None:
+        # The relationship is not the name of a field. Only create a relation.
+        logger.info(u'No field. Setting relation {} from {} to {}'.format(
+            source.absolute_url(), target.absolute_url(), relationship))
+        event._setRelation(source, from_attribute, RelationValue(to_id))
+        return
+
+    field, schema = field_and_schema
+
+    if isinstance(field, RelationList):
+        logger.info('Add relation to relationlist {} from {} to {}'.format(
+            from_attribute, source.absolute_url(), target.absolute_url()))
+        existing_relations = getattr(source, from_attribute, [])
+        existing_relations.append(RelationValue(to_id))
+        setattr(source, from_attribute, existing_relations)
+        modified(source)
+        return
+
+    elif isinstance(field, (Relation, RelationChoice)):
+        logger.info('Add relation {} from {} to {}'.format(
+            from_attribute, source.absolute_url(), target.absolute_url()))
+        setattr(source, from_attribute, RelationValue(to_id))
+        modified(source)
+        return
+
+    # We should never end up here!
+    logger.info('Warning: Unexpected relation {} from {} to {}'.format(
+        from_attribute, source.absolute_url(), target.absolute_url()))
+
+
+# @at_least_one_of('source', 'target', 'relationship')
```

Comment:
?

Reply:
Originally we had this restriction of needing at least one of source/target/relationship when deleting, but later I thought: when nothing is specified, we can delete all relations.
Might be a bit too implicit though. We can introduce a parameter `delete_all`. (`all` is a reserved keyword.) That seems nicer. I will do that.